### PR TITLE
Update extension with bundle changes

### LINKED ISSUE #74: Update extension to use new bundle format
# Problem

After https://github.com/jzaki/bls-wallet/pull/68 , the extension no longer works with the updated aggregator or smart contracts and the new bundle format for sending transactions to be aggregated in `bls-wallet-clients`.

# Solution

Update extension to use new bundle format in `bls-wallet-clients`.

### PR DESCRIPTION
**Pull request by jacque006**
## What is this PR doing?

- Updating extension to use new `bls-wallet-clients` and new bundle payload for sending transactions to be aggregated.
- Add PublicKey and PublicKeyStr to BLSWalletWrapper.
- Avoid publishing non-definition typescript files from typechain.
- Fix aggregator bundle POST path in clients.
- Remove @ethersproject specific deps from clients.
- Remove wallet creation paths from aggregator and extension.
- Update main README with client -> aggregator local workflow.
- Add send_eth.ts test script.
- Update labeler CI. Remove signer, don't add contracts on client only changes.

## How can these changes be manually tested?

Run extension and complete current onboarding workflow & successfully send a transaction to the aggregator.

## Does this PR resolve or contribute to any issues?

Resolves https://github.com/jzaki/bls-wallet/issues/74

## Checklist
- [x] I have manually tested these changes
- [x] Post a link to the PR in the group chat

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)

~TODO~
- [x] Remove wallet creation functionality from extension and aggregator.
- [x] Create issue for wallet creation to evaluate in future.

### CODE REVIEW

File: aggregator/deps.ts [hunk 1]
```diff
@@ -39,18 +39,18 @@ export type {
...
+} from "https://esm.sh/bls-wallet-clients@0.5.2";
 
 export {
   Aggregator as AggregatorClient,
   BlsWalletWrapper,
   getConfig,
   VerificationGateway,
   VerificationGateway__factory,
-} from "https://esm.sh/bls-wallet-clients@0.5.0-d6d7dd4";
+} from "https://esm.sh/bls-wallet-clients@0.5.2";
```
Comment (voltrevo) [inline aggregator/deps.ts]: Is there a reason you published non-experimental 0.5.2?

File: contracts/clients/package.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
   "license": "MIT",
   "private": false,
   "scripts": {
-    "build": "rm -rf dist && mkdir dist && cp -rH typechain dist/typechain && tsc",
+    "build": "rm -rf dist && mkdir dist && cp -rH typechain dist/typechain && find ./dist/typechain -type f \\! -name '*.d.ts' -name '*.ts' -delete && tsc",
```
Comment (voltrevo) [inline contracts/clients/package.json:12]: Avoids publishing non-definition typescript files from typechain? Nice. Could you add something about this to the PR description?

File: contracts/clients/tsconfig.json [hunk 3]
```diff
@@ -68,5 +68,5 @@
     /* Advanced Options */
     "skipLibCheck": true,                           /* Skip type checking of declaration files. */
     "forceConsistentCasingInFileNames": true        /* Disallow inconsistently-cased references to the same file. */
-  }
+  },
```
Comment (voltrevo) [inline contracts/clients/tsconfig.json]: Interesting trailing comma here. That's invalid json, but I suppose `tsc` is ok with it?

File: extension/source/App.ts [hunk 4]
```diff
@@ -192,7 +190,7 @@ export default class App {
       return undefined;
     }
 
-    return this.blsWalletSigner.getPublicKey(this.state.privateKey);
+    return this.wallet?.blsWalletSigner.getPublicKeyHash(this.state.privateKey);
```
Comment (voltrevo) [inline extension/source/App.ts]: Should be `getPublicKey` right? Not `getPublicKeyHash`?

(If the public key hash is what we really want then we should rename the method.)

File: extension/source/App.ts [hunk 5]
```diff
@@ -206,14 +204,9 @@ export default class App {
...
 
     try {
-      const creationTx = await BlsWallet.signCreation(
-        this.state.privateKey,
-        env.NETWORK_CONFIG.addresses.verificationGateway,
-        this.provider,
-      );
-
-      const createResult = await this.aggregator.createWallet(creationTx);
+      throw new Error('App: Wallet creation not implemented');
```
Comment (voltrevo) [inline extension/source/App.ts]: We're not doing an explicit create anymore, so the goal is to just remove this `createWallet` method entirely.

File: contracts/clients/package.json (hunk 2 above)
Comment (jacque006) [reply contracts/clients/package.json:12]: Added to description.

File: aggregator/deps.ts (hunk 1 above)
Comment (jacque006) [reply aggregator/deps.ts]: `0.5.0` failed webpack transpilation due to .ts files being included from typechain as you pointed out in https://github.com/jzaki/bls-wallet/pull/75#discussion_r776168168 . I thought I had a fix in `0.5.1` but realized shortly thereafter I had made a mistake. `0.5.2` has the .ts files properly removed.

Given that `0.4.x` -> `0.5.0` broke package level compatibility for the extension and potentially other similar consumers, it seemed prudent to push a prod tag resolving the issue.

File: extension/source/App.ts (hunk 4 above)
Comment (jacque006) [reply extension/source/App.ts]: I added this as typing filler, `getPublicKey` now returns an `Array(4)` instead of a string. This will be changed in future commits.

File: contracts/clients/tsconfig.json (hunk 3 above)
Comment (jacque006) [reply contracts/clients/tsconfig.json]: Apparently haha, will remove.

File: extension/source/App.ts (hunk 5 above)
Comment (jacque006) [reply extension/source/App.ts]: I still think there are cases where someone may want to non-lazily create a wallet, but I imagine those are not needed anytime soon.

I will add tasks to this PR to remove this, similar functionality in aggregator, and create an issue for manual wallet creation we can evaluate later.

File: extension/source/App.ts (hunk 4 above)
Comment (jacque006) [reply extension/source/App.ts]: Added `getPublicKeyStr` to signer and `PublicKey` and `PublicKeyStr` to `BLSWalletWrapper`.

File: extension/source/App.ts (hunk 5 above)
Comment (jacque006) [reply extension/source/App.ts]: Removed wallet creation paths and logic from agg and ext.

File: contracts/clients/package.json [hunk 6]
```diff
@@ -9,21 +9,15 @@
...
     "publish-experimental": "node scripts/showVersion.js >.version && npm version $(node scripts/showBaseVersion.js)-$(git rev-parse HEAD | head -c7) --allow-same-version && npm publish --tag experimental && npm version $(cat .version) && rm .version",
     "publish-experimental-dry-run": "node scripts/showVersion.js >.version && npm version $(node scripts/showBaseVersion.js)-$(git rev-parse HEAD | head -c7) --allow-same-version && npm publish --tag experimental --dry-run && npm version $(cat .version) && rm .version"
   },
   "dependencies": {
-    "@ethersproject/abi": "^5.4.0",
-    "@ethersproject/bignumber": "^5.4.1",
-    "@ethersproject/bytes": "^5.4.0",
-    "@ethersproject/keccak256": "^5.4.0",
-    "@ethersproject/sha2": "^5.4.0",
-    "@ethersproject/solidity": "^5.4.0",
```
Comment (jacque006) [inline contracts/clients/package.json:25]: @voltrevo Given discussion today and what I saw in my tests this seemed safe to do. Let me know if I should revert.

File: extension/source/Background/index.ts [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
 import RequestHandler from './RequestHandler';
 
-import { AGGREGATOR_URL, NETWORK_CONFIG, CHAIN_RPC_URL } from '../env';
+import { AGGREGATOR_URL /* , NETWORK_CONFIG */, CHAIN_RPC_URL } from '../env';
 
 (async () => {
   const app = new App(
-    await initBlsWalletSigner({ chainId: NETWORK_CONFIG.auxiliary.chainid }),
+    // TODO Is this still needed for something else?
+    // await initBlsWalletSigner({ chainId: NETWORK_CONFIG.auxiliary.chainid }),
```
Comment (jacque006) [inline extension/source/Background/index.ts]: @voltrevo @jzaki Due to `BLSWalletWrapper` now wrapping the signer, this does not appear to be needed anymore. Safe to remove?

File: extension/source/Popup/index.tsx [hunk 8]
```diff
@@ -1,19 +1,19 @@
...
+import { AGGREGATOR_URL /* , NETWORK_CONFIG */, CHAIN_RPC_URL } from '../env';
 import Popup from './Popup';
 
 import './styles.scss';
 
 const appPromise = (async () =>
   new App(
-    await initBlsWalletSigner({ chainId: NETWORK_CONFIG.auxiliary.chainid }),
+    // TODO Is this still needed somewhere else?
+    // await initBlsWalletSigner({ chainId: NETWORK_CONFIG.auxiliary.chainid }),
```
Comment (jacque006) [inline extension/source/Popup/index.tsx]: @voltrevo @jzaki Same here.

**Comment by jacque006** [comment]
@voltrevo @kautukkundan Feel free to push any changes to this if you see something that needs to be corrected.

File: extension/source/App.ts (hunk 4 above)
Comment (voltrevo) [reply extension/source/App.ts]: Interesting. I want to aim for handling these things in a more uniform way but not sure whether it's suitable here or not. In any case, it's something I'd rather fix as a followup than have in this PR if it looks like a good idea later.

Comment (voltrevo) [reply extension/source/App.ts]: ("Handling in a uniform way" means using the string tuple everywhere instead of converting between formats for different projects.)

File: contracts/clients/package.json (hunk 6 above)
Comment (voltrevo) [reply contracts/clients/package.json:25]: Nah, looks good. Nice one.

File: extension/source/App.ts (hunk 5 above)
Comment (voltrevo) [reply extension/source/App.ts]: > I still think there are cases where someone may want to non-lazily create a wallet, but I imagine those are not needed anytime soon.

True. Should be ok with just submitting an empty operation from the client for that though. Keeps the interface simple.

File: extension/source/Background/index.ts (hunk 7 above)
Comment (voltrevo) [reply extension/source/Background/index.ts]: Yeah that makes sense. Safe to remove 👍.

File: extension/source/Popup/index.tsx (hunk 8 above)
Comment (voltrevo) [reply extension/source/Popup/index.tsx]: Safe to remove 👍.

File: contracts/clients/src/signer/getPublicKeyStr.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    signerFactory: hubbleBls.signer.BlsSignerFactory,
+    domain: Uint8Array,
+  ) =>
+  (privateKey: string): string => {
+    const [x0, x1, y0, y1] = getPublicKey(signerFactory, domain)(privateKey);
+    return hubbleBls.mcl.dumpG2([
+      x0.toString(),
+      x1.toString(),
+      y0.toString(),
+      y1.toString(),
```
Comment (voltrevo) [inline contracts/clients/src/signer/getPublicKeyStr.ts:14]: The `.toString()`s are a bit odd here... not sure if you realized. I want to explain what's going on here.

The result of `getPublicKey` is understandably typed as `PublicKey`, which is the wide/permissive type we get from typechain, so the elements are `BigNumberish`.

`dumpG2` requires strings, and so `.toString()`, being available on all possibilities of `BigNumberish` satisfies the type checker. However, `dumpG2` actually expects _hex_ strings (right?), so this will only work if all the `BigNumberish` elements are actually already hex strings.

...which they are, because they come from `signer.pubkey`.

tl;dr - `.toString()` satisfies the type checker, but the potential issue flagged by the type checker isn't actually fixed by `.toString()`.

I'm nit picking. I think. It's unlikely this will ever be an issue, but it's a cause of confusion and could be the cause of a bug with subsequent changes that would look reasonable.

I think I'll submit a followup PR that replaces the use of `getPublicKey` with the underlying hubblebls call. Using `as string` would also be preferable, since at least it keeps the flag that something odd is going on (because it is), and avoids the no-op.

Comment (voltrevo) [reply contracts/clients/src/signer/getPublicKeyStr.ts:14]: https://github.com/jzaki/bls-wallet/issues/83

Comment (voltrevo) [reply contracts/clients/src/signer/getPublicKeyStr.ts:14]: (Btw it's very possible that I also used the same kind of `.toString()` in the past 😅.)

File: extension/source/Background/index.ts (hunk 7 above)
Comment (voltrevo) [reply extension/source/Background/index.ts]:
```suggestion
```

File: extension/source/Popup/index.tsx [hunk 10]
```diff
@@ -1,19 +1,19 @@
-import { Aggregator } from 'bls-wallet-clients';
-import { initBlsWalletSigner } from 'bls-wallet-signer';
+import { Aggregator /* , initBlsWalletSigner */ } from 'bls-wallet-clients';
```
Comment (voltrevo) [inline extension/source/Popup/index.tsx]:
```suggestion
import { Aggregator } from 'bls-wallet-clients';
```

File: extension/source/Popup/index.tsx [hunk 11]
```diff
@@ -1,19 +1,19 @@
...
 import { ethers } from 'ethers';
 import * as React from 'react';
 import ReactDOM from 'react-dom';
 import { browser } from 'webextension-polyfill-ts';
 
 import App from '../App';
-import { AGGREGATOR_URL, NETWORK_CONFIG, CHAIN_RPC_URL } from '../env';
+import { AGGREGATOR_URL /* , NETWORK_CONFIG */, CHAIN_RPC_URL } from '../env';
```
Comment (voltrevo) [inline extension/source/Popup/index.tsx]:
```suggestion
import { AGGREGATOR_URL, CHAIN_RPC_URL } from '../env';
```

File: extension/source/Popup/index.tsx (hunk 8 above)
Comment (voltrevo) [reply extension/source/Popup/index.tsx]:
```suggestion
```

**Review by voltrevo — APPROVED** [review]
Amazing stuff 👍

File: contracts/clients/src/signer/getPublicKeyStr.ts (hunk 9 above)
Comment (jacque006) [reply contracts/clients/src/signer/getPublicKeyStr.ts:14]: > However, dumpG2 actually expects hex strings (right?), so this will only work if all the BigNumberish elements are actually already hex strings. ...which they are, because they come from signer.pubkey.

(dumpG2)[https://github.com/jzaki/bls-wallet/blob/e366d0af4c339db3f380732550c9319ea4d6982b/contracts/clients/deps/hubble-bls/mcl.ts#L223-L226] explicitly needs the pubkey segments to be strings, and I imagine in hex format. Good eye, there are some cases where this will fail spectacularly if it is a `number` or `BigNumber`:
Screenshot from 2022-01-04 14-01-37

We might want to just have the type for our pubkey be `[string, string,string, string]`, and then we can have a function that converts from the typechain type to our needed type (I don't think `hubble-bls` exports `solG2`). We could also just make a seperate wrapper class for it and put needed utility functions in that class similar to https://github.com/thehubbleproject/hubble-contracts/blob/master/ts/pubkey.ts

Comment (jzaki) [reply contracts/clients/src/signer/getPublicKeyStr.ts:14]: Perhaps relevant here, some conversions remove preceding 0's from hex, which is problematic when a smart contract call requires the right length. Eg a 20 byte address:
`0x0102030405060708091011121314151617181920`, becoming,
`0x102030405060708091011121314151617181920`
